File: src/docker_volume_tools/backup.py
"""Module for backing up Docker volumes."""

import os
import json
import datetime
import tempfile
from typing import Dict, List, Optional
import docker
from docker.models.volumes import Volume
from .compose import VolumeInfo, parse_compose_file

class BackupError(Exception):
    """Error during backup operation."""
    pass

def create_backup(
    compose_file: str,
    output_dir: str,
    compress: bool = True,
    volumes_to_backup: Optional[List[str]] = None,
    ssh_target: Optional[str] = None
) -> str:
    """Create a backup of Docker volumes from a compose project.
    
    Args:
        compose_file: Path to the docker-compose.yml file
        output_dir: Directory to store the backup
        compress: Whether to compress the backup
        volumes_to_backup: List of volume names to backup (None for all)
        ssh_target: Optional SSH target for direct transfer (format: user@host:path)
    
    Returns:
        Path to the backup file or SSH target path if using direct transfer
    """
    # Parse compose file
    volumes = parse_compose_file(compose_file)
    print("\nParsed volumes:")
    for v in volumes:
        print(f"- {v.name} -> {v.compose_name} (type: {v.type})")
    
    # Filter volumes if needed
    if volumes_to_backup:
        volumes = [v for v in volumes if v.name in volumes_to_backup]
        
    if not volumes:
        raise BackupError("No volumes to backup")
    
    # Filter out bind mounts and deduplicate volumes by name
    volume_dict = {}
    for volume in volumes:
        if volume.type == 'named':
            # Keep only the first occurrence of each volume (primary service)
            if volume.name not in volume_dict:
                volume_dict[volume.name] = volume
    
    named_volumes = list(volume_dict.values())
    if not named_volumes:
        raise BackupError("No named volumes to backup")
        
    print("\nVolumes to backup:")
    for v in named_volumes:
        print(f"- {v.name} -> {v.compose_name}")
    
    # Create Docker client
    client = docker.from_env()
    
    # Generate backup name
    timestamp = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
    project_name = os.path.splitext(os.path.basename(compose_file))[0]
    backup_name = f"{project_name}_volumes_{timestamp}"
    
    # Prepare metadata
    metadata = {
        'timestamp': timestamp,
        'project': project_name,
        'compose_file': os.path.basename(compose_file),
        'volumes': [
            {
                'name': v.compose_name,
                'service': v.service,
                'target': v.target,
                'is_external': v.is_external
            }
            for v in named_volumes
        ]
    }
    
    # Create temporary directory for metadata and SSH
    with tempfile.TemporaryDirectory() as temp_dir:
        metadata_file = os.path.join(temp_dir, 'metadata.json')
        with open(metadata_file, 'w') as f:
            json.dump(metadata, f, indent=2)
        
        # Prepare volume mount options for backup container
        volume_mounts = {
            v.compose_name: {'bind': f'/volumes/{v.name}', 'mode': 'ro'}
            for v in named_volumes
        }
        volume_mounts[metadata_file] = {'bind': '/metadata.json', 'mode': 'ro'}
        
        if ssh_target:
            # Direct SSH transfer
            ssh_user_host, ssh_path = ssh_target.split(':', 1)
            remote_file = f"{ssh_path}/{backup_name}.{'tar.gz' if compress else 'tar'}"
            
            # Prepare SSH environment
            ssh_dir = os.path.join(temp_dir, '.ssh')
            os.makedirs(ssh_dir, mode=0o700)
            
            # Copy SSH config and keys
            home = os.path.expanduser("~")
            user_ssh_dir = os.path.join(home, '.ssh')
            
            # Copy known_hosts if exists
            known_hosts = os.path.join(user_ssh_dir, 'known_hosts')
            if os.path.exists(known_hosts):
                os.system(f"cp {known_hosts} {ssh_dir}/")
            
            # Copy SSH keys
            for key_file in ['id_rsa', 'id_ed25519']:
                key_path = os.path.join(user_ssh_dir, key_file)
                if os.path.exists(key_path):
                    os.system(f"cp {key_path} {ssh_dir}/")
                    os.chmod(os.path.join(ssh_dir, key_file), 0o600)
            
            # Add SSH directory to volume mounts (en lecture/écriture pour known_hosts)
            volume_mounts[ssh_dir] = {'bind': '/root/.ssh', 'mode': 'rw'}
            
            try:
                # Create and start container with SSH support
                container = client.containers.run(
                    'alpine:latest',
                    command="sleep infinity",  # Keep container running
                    volumes=volume_mounts,
                    network_mode="host",  # Important pour SSH
                    detach=True  # Run in background
                )
                
                # Install SSH in container
                print("Installing SSH client in container...")
                result = container.exec_run("apk add --no-cache openssh-client")
                if result.exit_code != 0:
                    raise BackupError(f"Failed to install SSH client: {result.output.decode()}")
                
                # Create remote directory if it doesn't exist
                print(f"Creating remote directory {ssh_path}...")
                mkdir_cmd = f"ssh -o StrictHostKeyChecking=accept-new {ssh_user_host} 'mkdir -p {ssh_path}'"
                result = container.exec_run(["/bin/sh", "-c", mkdir_cmd])
                if result.exit_code != 0:
                    raise BackupError(f"Failed to create remote directory: {result.output.decode()}")
                
                # Construire la commande tar pour créer une archive avec tous les volumes
                print("Starting backup transfer...")
                
                # Créer un répertoire temporaire pour organiser les volumes
                result = container.exec_run("mkdir -p /tmp_backup")
                if result.exit_code != 0:
                    raise BackupError("Failed to create temporary directory")
                
                # Copier le metadata
                result = container.exec_run("cp /metadata.json /tmp_backup/")
                if result.exit_code != 0:
                    raise BackupError("Failed to copy metadata")
                
                # Créer des liens symboliques vers les volumes dans le répertoire temporaire
                result = container.exec_run("mkdir -p /tmp_backup/volumes")
                if result.exit_code != 0:
                    raise BackupError("Failed to create volumes directory")
                
                # Créer des liens symboliques vers chaque volume
                for v in named_volumes:
                    print(f"Adding volume {v.name} to backup...")
                    result = container.exec_run(f"ln -s /volumes/{v.name} /tmp_backup/volumes/{v.name}")
                    if result.exit_code != 0:
                        raise BackupError(f"Failed to create symlink for volume {v.name}")
                
                # Créer l'archive tar avec tous les fichiers
                # Utiliser -h pour suivre les liens symboliques
                tar_cmd = "cd /tmp_backup && tar -hcf - ."
                if compress:
                    tar_cmd = f"({tar_cmd}) | gzip"
                
                # Pipe vers SSH
                ssh_cmd = f"ssh -o StrictHostKeyChecking=accept-new {ssh_user_host} 'cat > {remote_file}'"
                full_cmd = f"{tar_cmd} | {ssh_cmd}"
                
                # Exécuter la commande
                result = container.exec_run(
                    ["/bin/sh", "-c", full_cmd],
                    demux=True  # Separate stdout/stderr
                )
                
                if result.exit_code != 0:
                    error_msg = result.output[1].decode() if result.output[1] else "Unknown error"
                    raise BackupError(f"Backup transfer failed: {error_msg}")
                
                # Nettoyer le répertoire temporaire
                container.exec_run("rm -rf /tmp_backup")
                
                return remote_file
                
            except Exception as e:
                print(f"Error during SSH transfer: {str(e)}")
                raise BackupError(f"Failed to transfer backup via SSH: {str(e)}")
            finally:
                # Cleanup
                try:
                    container.stop()
                    container.remove()
                except:
                    pass
        else:
            # Local backup
            if not os.path.exists(output_dir):
                os.makedirs(output_dir)
            
            backup_dir = os.path.join(output_dir, backup_name)
            os.makedirs(backup_dir)
            
            # Backup each volume locally
            for volume in named_volumes:
                print(f"\nBacking up volume {volume.compose_name}...")
                tar_name = f"{volume.name}.{'tar.gz' if compress else 'tar'}"
                tar_cmd = f"tar {'cz' if compress else 'c'}f /backup/{tar_name} -C /volumes/{volume.name} ."
                
                try:
                    backup_container = client.containers.run(
                        'alpine:latest',
                        command=tar_cmd,
                        volumes={
                            volume.compose_name: {'bind': f'/volumes/{volume.name}', 'mode': 'ro'},
                            backup_dir: {'bind': '/backup', 'mode': 'rw'}
                        },
                        remove=True
                    )
                except Exception as e:
                    print(f"Error backing up volume {volume.compose_name}: {str(e)}")
                    raise BackupError(f"Failed to backup volume {volume.compose_name}: {str(e)}")
            
            # Copy metadata
            os.system(f"cp {metadata_file} {backup_dir}/metadata.json")
            
            # Create final archive
            final_archive = os.path.join(output_dir, f"{backup_name}.{'tar.gz' if compress else 'tar'}")
            archive_cmd = f"tar {'cz' if compress else 'c'}f {final_archive} -C {output_dir} {backup_name}"
            os.system(archive_cmd)
            
            # Cleanup temporary directory
            os.system(f"rm -rf {backup_dir}")
            
            return final_archive

def get_volume_size(volume: Volume) -> str:
    """Get the size of a Docker volume.
    
    Args:
        volume: Docker volume object
        
    Returns:
        Human readable size string
    """
    try:
        # Create temporary container to check size
        container = docker.from_env().containers.run(
            'alpine:latest',
            command=f"du -sh /volume",
            volumes={volume.name: {'bind': '/volume', 'mode': 'ro'}},
            remove=True
        )
        size = container.decode().split()[0]
        return size
    except Exception:
        return "unknown" 

File: internal/backup/backup.go
package backup

import (
	"archive/tar"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/pbouamriou/docker-volume-tools/internal/compose"
	"github.com/pbouamriou/docker-volume-tools/internal/docker"
	"github.com/pbouamriou/docker-volume-tools/pkg/models"
)

// CreateBackup creates a backup of Docker volumes from a compose project
func CreateBackup(options models.BackupOptions) (string, error) {
	// Parse compose file
	volumes, err := compose.ParseComposeFile(options.ComposeFile)
	if err != nil {
		return "", fmt.Errorf("failed to parse compose file: %w", err)
	}

	// Filter volumes if specified
	if len(options.VolumesToBackup) > 0 {
		volumes = filterVolumes(volumes, options.VolumesToBackup)
	}

	// Filter only named volumes and deduplicate
	namedVolumes := deduplicateVolumes(filterNamedVolumes(volumes))
	if len(namedVolumes) == 0 {
		return "", fmt.Errorf("no named volumes to backup")
	}

	// Create Docker client
	dockerClient, err := docker.NewClient()
	if err != nil {
		return "", err
	}
	defer dockerClient.Close()

	// Generate backup name
	timestamp := models.GenerateTimestamp()
	projectName := filepath.Base(strings.TrimSuffix(options.ComposeFile, filepath.Ext(options.ComposeFile)))
	backupName := fmt.Sprintf("%s_volumes_%s", projectName, timestamp)

	// Create backup directory or use SSH
	if options.SSHTarget != "" {
		return createSSHBackup(dockerClient, namedVolumes, backupName, projectName, options)
	}

	return createLocalBackup(dockerClient, namedVolumes, backupName, projectName, options)
}

// createLocalBackup creates a local backup archive
func createLocalBackup(dockerClient *docker.Client, volumes []models.VolumeInfo, backupName, projectName string, options models.BackupOptions) (string, error) {
	// Create output directory
	if err := os.MkdirAll(options.OutputDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create output directory: %w", err)
	}

	// Create temporary directory for backup
	tempDir := filepath.Join(options.OutputDir, backupName)
	if err := os.MkdirAll(tempDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create temp directory: %w", err)
	}
	defer os.RemoveAll(tempDir)

	// Create volumes subdirectory
	volumesDir := filepath.Join(tempDir, "volumes")
	if err := os.MkdirAll(volumesDir, 0755); err != nil {
		return "", fmt.Errorf("failed to create volumes directory: %w", err)
	}

	// Backup each volume
	var volumeBackups []models.VolumeBackup
	for _, vol := range volumes {
		fmt.Printf("Backing up volume %s...\n", vol.ComposeName)

		volumeBackupDir := filepath.Join(volumesDir, vol.Name)
		if err := os.MkdirAll(volumeBackupDir, 0755); err != nil {
			return "", fmt.Errorf("failed to create volume backup dir: %w", err)
		}

		if err := backupVolume(dockerClient, vol.ComposeName, volumeBackupDir); err != nil {
			return "", fmt.Errorf("failed to backup volume %s: %w", vol.ComposeName, err)
		}

		volumeBackups = append(volumeBackups, models.VolumeBackup{
			Name:        vol.ComposeName,
			Service:     vol.Service,
			Target:      vol.Target,
			IsExternal:  vol.IsExternal,
			ArchivePath: vol.Name,
		})
	}

	// Create metadata
	metadata := models.BackupMetadata{
		Timestamp:   models.GenerateTimestamp(),
		Project:     projectName,
		ComposeFile: filepath.Base(options.ComposeFile),
		Volumes:     volumeBackups,
	}

	metadataBytes, err := json.MarshalIndent(metadata, "", "  ")
	if err != nil {
		return "", fmt.Errorf("failed to marshal metadata: %w", err)
	}

	metadataPath := filepath.Join(tempDir, "metadata.json")
	if err := os.WriteFile(metadataPath, metadataBytes, 0644); err != nil {
		return "", fmt.Errorf("failed to write metadata: %w", err)
	}

	// Create final archive
	ext := ".tar"
	if options.Compress {
		ext = ".tar.gz"
	}
	archivePath := filepath.Join(options.OutputDir, backupName+ext)

	if err := createTarArchive(tempDir, archivePath, options.Compress); err != nil {
		return "", fmt.Errorf("failed to create archive: %w", err)
	}

	return archivePath, nil
}

// backupVolume backs up a single Docker volume to a directory
// IMPORTANT: This preserves symlinks (does NOT dereference them)
func backupVolume(dockerClient *docker.Client, volumeName, outputDir string) error {
	// Pull alpine image
	if err := dockerClient.PullImage("alpine:latest"); err != nil {
		return fmt.Errorf("failed to pull alpine image: %w", err)
	}

	// Create temporary container with volume mounted
	containerID, err := dockerClient.RunContainer(docker.ContainerConfig{
		Image: "alpine:latest",
		Name:  fmt.Sprintf("backup_%s", volumeName),
		Cmd:   []string{"sleep", "infinity"},
		Volumes: map[string]string{
			volumeName: "/volume",
		},
	})
	if err != nil {
		return err
	}
	defer dockerClient.Cleanup(containerID)

	// Copy files from container preserving symlinks
	// Use docker cp which preserves symlinks by default
	copyCmd := fmt.Sprintf("docker cp %s:/volume/. %s/", containerID, outputDir)
	if err := runCommand(copyCmd); err != nil {
		return fmt.Errorf("failed to copy volume data: %w", err)
	}

	return nil
}

// createTarArchive creates a tar archive from a directory
// CRITICAL: Preserves symlinks by detecting and storing them correctly
func createTarArchive(sourceDir, targetPath string, compress bool) error {
	// Create output file
	outFile, err := os.Create(targetPath)
	if err != nil {
		return fmt.Errorf("failed to create archive file: %w", err)
	}
	defer outFile.Close()

	// Setup compression if needed
	var writer io.Writer = outFile
	if compress {
		gzWriter := gzip.NewWriter(outFile)
		defer gzWriter.Close()
		writer = gzWriter
	}

	// Create tar writer
	tarWriter := tar.NewWriter(writer)
	defer tarWriter.Close()

	// Walk directory and add files
	return filepath.Walk(sourceDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Get relative path
		relPath, err := filepath.Rel(sourceDir, path)
		if err != nil {
			return err
		}

		// Skip root directory
		if relPath == "." {
			return nil
		}

		// Create tar header
		header, err := tar.FileInfoHeader(info, "")
		if err != nil {
			return fmt.Errorf("failed to create header for %s: %w", path, err)
		}
		header.Name = relPath

		// Handle symlinks specially
		if info.Mode()&os.ModeSymlink != 0 {
			// Read the symlink target
			linkTarget, err := os.Readlink(path)
			if err != nil {
				return fmt.Errorf("failed to read symlink %s: %w", path, err)
			}
			header.Typeflag = tar.TypeSymlink
			header.Linkname = linkTarget
			header.Size = 0

			// Write header only (no file content for symlinks)
			if err := tarWriter.WriteHeader(header); err != nil {
				return fmt.Errorf("failed to write symlink header: %w", err)
			}
			fmt.Printf("  Preserved symlink: %s -> %s\n", relPath, linkTarget)
			return nil
		}

		// Write header
		if err := tarWriter.WriteHeader(header); err != nil {
			return fmt.Errorf("failed to write header: %w", err)
		}

		// If it's a file, write its content
		if info.Mode().IsRegular() {
			file, err := os.Open(path)
			if err != nil {
				return fmt.Errorf("failed to open file %s: %w", path, err)
			}
			defer file.Close()

			if _, err := io.Copy(tarWriter, file); err != nil {
				return fmt.Errorf("failed to write file %s: %w", path, err)
			}
		}

		return nil
	})
}

// Helper functions

func filterVolumes(volumes []models.VolumeInfo, names []string) []models.VolumeInfo {
	nameMap := make(map[string]bool)
	for _, name := range names {
		nameMap[name] = true
	}

	var filtered []models.VolumeInfo
	for _, vol := range volumes {
		if nameMap[vol.Name] {
			filtered = append(filtered, vol)
		}
	}
	return filtered
}

func filterNamedVolumes(volumes []models.VolumeInfo) []models.VolumeInfo {
	var named []models.VolumeInfo
	for _, vol := range volumes {
		if vol.Type == "named" {
			named = append(named, vol)
		}
	}
	return named
}

func deduplicateVolumes(volumes []models.VolumeInfo) []models.VolumeInfo {
	seen := make(map[string]bool)
	var deduplicated []models.VolumeInfo

	for _, vol := range volumes {
		if !seen[vol.Name] {
			seen[vol.Name] = true
			deduplicated = append(deduplicated, vol)
		}
	}
	return deduplicated
}

func runCommand(cmdStr string) error {
	parts := strings.Fields(cmdStr)
	if len(parts) == 0 {
		return fmt.Errorf("empty command")
	}

	cmd := exec.Command(parts[0], parts[1:]...)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("command failed: %s: %w", string(output), err)
	}
	return nil
}

// createSSHBackup creates a backup and streams it directly via SSH without using local disk space
// This method uses a Docker container to mount all volumes and stream the tar archive directly to SSH
func createSSHBackup(dockerClient *docker.Client, volumes []models.VolumeInfo, backupName, projectName string, options models.BackupOptions) (string, error) {
	// Parse SSH target (format: user@host:path)
	parts := strings.Split(options.SSHTarget, ":")
	if len(parts) != 2 {
		return "", fmt.Errorf("invalid SSH target format, expected user@host:path")
	}

	userHost := parts[0]
	remotePath := parts[1]

	// Determine file extension
	ext := ".tar"
	if options.Compress {
		ext = ".tar.gz"
	}
	remoteFile := fmt.Sprintf("%s/%s%s", remotePath, backupName, ext)

	// Prepare metadata
	var volumeBackups []models.VolumeBackup
	for _, vol := range volumes {
		volumeBackups = append(volumeBackups, models.VolumeBackup{
			Name:        vol.ComposeName,
			Service:     vol.Service,
			Target:      vol.Target,
			IsExternal:  vol.IsExternal,
			ArchivePath: vol.Name,
		})
	}

	metadata := models.BackupMetadata{
		Timestamp:   models.GenerateTimestamp(),
		Project:     projectName,
		ComposeFile: filepath.Base(options.ComposeFile),
		Volumes:     volumeBackups,
	}

	metadataBytes, err := json.MarshalIndent(metadata, "", "  ")
	if err != nil {
		return "", fmt.Errorf("failed to marshal metadata: %w", err)
	}

	// Create a small temp directory ONLY for metadata and SSH config (very small files)
	tempDir, err := os.MkdirTemp("", "dvtools-metadata-*")
	if err != nil {
		return "", fmt.Errorf("failed to create temp dir: %w", err)
	}
	defer os.RemoveAll(tempDir)

	metadataPath := filepath.Join(tempDir, "metadata.json")
	if err := os.WriteFile(metadataPath, metadataBytes, 0644); err != nil {
		return "", fmt.Errorf("failed to write metadata: %w", err)
	}

	// Copy SSH config to temp directory
	sshDir := filepath.Join(tempDir, ".ssh")
	if err := os.MkdirAll(sshDir, 0700); err != nil {
		return "", fmt.Errorf("failed to create SSH directory: %w", err)
	}

	homeDir, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("failed to get home directory: %w", err)
	}

	userSSHDir := filepath.Join(homeDir, ".ssh")

	// Copy known_hosts if exists
	knownHosts := filepath.Join(userSSHDir, "known_hosts")
	if _, err := os.Stat(knownHosts); err == nil {
		exec.Command("cp", knownHosts, sshDir).Run()
	}

	// Copy SSH keys
	for _, keyFile := range []string{"id_rsa", "id_ed25519"} {
		keyPath := filepath.Join(userSSHDir, keyFile)
		if _, err := os.Stat(keyPath); err == nil {
			destPath := filepath.Join(sshDir, keyFile)
			exec.Command("cp", keyPath, destPath).Run()
			os.Chmod(destPath, 0600)
		}
	}

	// Prepare volume mounts for container
	volumeMounts := make(map[string]string)
	for _, vol := range volumes {
		volumeMounts[vol.ComposeName] = fmt.Sprintf("/volumes/%s", vol.Name)
	}

	// Add metadata and SSH directory as bind mounts (not volume mounts)
	// Note: The RunContainer method will handle these correctly
	volumeMounts[metadataPath] = "/metadata.json"
	volumeMounts[sshDir] = "/root/.ssh"

	// Create and start container
	fmt.Println("Creating temporary container for streaming backup...")
	containerID, err := dockerClient.RunContainer(docker.ContainerConfig{
		Image:       "alpine:latest",
		Name:        "",
		Cmd:         []string{"sleep", "infinity"},
		Volumes:     volumeMounts,
		NetworkMode: "host", // Important for SSH
	})
	if err != nil {
		return "", fmt.Errorf("failed to create container: %w", err)
	}
	defer dockerClient.Cleanup(containerID)

	// Install SSH client in container
	fmt.Println("Installing SSH client in container...")
	result, err := dockerClient.ExecInContainer(containerID, []string{"apk", "add", "--no-cache", "openssh-client"})
	if err != nil || result.ExitCode != 0 {
		return "", fmt.Errorf("failed to install SSH client: %v %s", err, result.Stderr)
	}

	// Create remote directory
	fmt.Printf("Creating remote directory %s...\n", remotePath)
	mkdirCmd := fmt.Sprintf("ssh -o StrictHostKeyChecking=accept-new %s 'mkdir -p %s'", userHost, remotePath)
	result, err = dockerClient.ExecInContainer(containerID, []string{"/bin/sh", "-c", mkdirCmd})
	if err != nil || result.ExitCode != 0 {
		return "", fmt.Errorf("failed to create remote directory: %v %s", err, result.Stderr)
	}

	// Prepare temporary backup structure inside container
	fmt.Println("Starting backup transfer...")

	commands := []string{
		"mkdir -p /tmp_backup/volumes",
		"cp /metadata.json /tmp_backup/",
	}

	// Create symbolic links to volumes
	for _, vol := range volumes {
		cmd := fmt.Sprintf("ln -s /volumes/%s /tmp_backup/volumes/%s", vol.Name, vol.Name)
		commands = append(commands, cmd)
	}

	// Execute setup commands
	for _, cmd := range commands {
		result, err = dockerClient.ExecInContainer(containerID, []string{"/bin/sh", "-c", cmd})
		if err != nil || result.ExitCode != 0 {
			return "", fmt.Errorf("failed to execute command %s: %v %s", cmd, err, result.Stderr)
		}
	}

	// Stream tar archive directly to SSH
	// Use -h flag to follow symlinks (dereference them)
	// IMPORTANT: We tar metadata.json first, then volumes, to ensure fast validation
	// (avoids having to read the entire archive to find metadata at the end)
	tarCmd := "cd /tmp_backup && tar -hcf - metadata.json volumes"
	if options.Compress {
		tarCmd = fmt.Sprintf("(%s) | gzip", tarCmd)
	}

	sshCmd := fmt.Sprintf("ssh -o StrictHostKeyChecking=accept-new %s 'cat > %s'", userHost, remoteFile)
	fullCmd := fmt.Sprintf("%s | %s", tarCmd, sshCmd)

	fmt.Printf("Streaming backup to %s...\n", options.SSHTarget)
	result, err = dockerClient.ExecInContainer(containerID, []string{"/bin/sh", "-c", fullCmd})
	if err != nil || result.ExitCode != 0 {
		return "", fmt.Errorf("backup transfer failed: %v %s", err, result.Stderr)
	}

	// Cleanup temp directory in container
	dockerClient.ExecInContainer(containerID, []string{"rm", "-rf", "/tmp_backup"})

	fmt.Println("Transfer completed successfully!")
	return remoteFile, nil
}
